Update PL_main_thread on fork() and add more tests

tonycoz opened the pull request:
<!--
A good description should explain the problem the pull request addresses
and give context to the reviewers to aid them in their reviews.
-->
Ensure the saved PL_main_thread when the process forks, this ensures we are still forwarding signals received by non-perl threads to the main thread.

This turns out not to be necessary on Linux and FreeBSD as tested, since the main thread id in parent and child processes appear to have the same value (they compare equal with pthread_equal()), but this may not be true for other platforms and may change for Linux or FreeBSD.

Also add extra on-point tests for the original issue from #22487

<!--
Significant changes to Perl must be documented in perldelta.

Consider if the changes in this pull request are worthy of a perldelta
entry, then pick the appropriate line below and remove the others.
-->

Not sure if this needs a perldelta.

---------------------------------------------------------------------------------
* This set of changes requires a perldelta entry, and I need help writing it.

tonycoz committed:
update PL_main_thread on fork()

85e97066 modified the perl signal handler to forward signals to the
main thread if it received a signal in a non-perl thread, which
required saving the id of the main perl thread.

Unfortunately I forgot to handle a possible change in the main thread
id on a fork, this fixes that by re-saving the new main thread id
immediately after a fork (via pthread_atfork())

On Linux it appears that the main thread id returned by pthread_seld()
is constant between processes, but this may not be true on other
platforms.

Discussed at:

https://github.com/Perl/perl5/issues/23326#issuecomment-3050481975

tonycoz committed:
add direct tests for the bug reported in #22487

This was fixed by 85e97066 but the test only checked the sanity of the
saved main thread thread id.

Tested locally for failure by disarming the change in
Perl_csighandler3.

bulk88 commented on ext/XS-APItest/APItest.xs:
```diff
@@ -4367,6 +4378,21 @@ CODE:
+make_signal_thread()
+CODE:
+    if (pthread_create(&RETVAL, NULL, signal_thread_start, NULL) != 0)
+        XSRETURN_EMPTY;
```
Why return empty list instead of `&PL_sv_undef` or `&PL_sv_no` or SvPV "" empty str or less perfect, SvIV -1 or SvIV 0? I know `pp_entersub` will cast empty list to 1 x `&PL_sv_undef`  in `G_SCALAR` cxt, but why mk `pp_entersub do the cast?

bulk88 commented on ext/XS-APItest/t/thread.t:
```diff
@@ -1,17 +1,71 @@
+    else {
+        is(waitpid($pid, 0), $pid, "wait child");
+        # catches the child segfaulting for example
+        is($?, 0, "child success");
```
would `test.pl`/`Test2::` `sub is()` corrupt  PP mg var `$?`  between `waitpid()` ret and sampling of `$?`?

tonycoz replied:
There was some discussion of this during Test2 integration - it should be preserving special variables like `$!`, `$^E`, `$?` and I assume it does.

bulk88 commented on ext/XS-APItest/typemap:
```diff
@@ -3,12 +3,26 @@ XS::APItest::PtrTable		T_PTROBJ
         sv_setpvn($arg, (const char *)($var), sizeof(WCHAR) * (1+wcslen($var)));
+
+T_THREADID
+        sv_setpvn($arg, (const char *)&($var), sizeof($var));
```
I see paranoia above with `SvPVbyte()`, so `OUTPUT:` also needs paranoia. `sv_setpvn()` doesn't clear the UTF8 flag if TARG already has UTF8 flag on. Also following paranoia guidelines, you forgot to call a `_mg()` `SV*` setter variant. Per rules, remember to clear UTF8 flag before the SMG setter method fires, not after.  It should be `sv_setpvn(); SvUTF8_off(); SvSETMAGIC();` or better yet `SvUTF8_off(); sv_setpvn_mg();`.

I am very aware this is code is inside XS-APItest, and not some high river CPAN XS module, so XS  code "quality" is a very low priority here, but I saw paranoia INPUT: type entry so I'll keep the paranoia going.

tonycoz replied:
You don't need to use sv_setpvn_mg().

XS will generate a SvSETMAGIC() if needed, eg if I change to:
```
int
make_signal_thread2(OUT pthread_t id)
CODE:
    RETVAL = pthread_create(&id, NULL, signal_thread_start, NULL) != 0;
OUTPUT:
    RETVAL
```
(int to avoid the boolSV() optimization)

The code generated is:
```
XS_EUPXS(XS_XS__APItest_make_signal_thread2)
{
    dVAR; dXSARGS;
    if (items != 1)
       croak_xs_usage(cv,  "id");
    {
        int     RETVAL;
        dXSTARG;
        pthread_t       id;
#line 4392 "APItest.xs"
    RETVAL = pthread_create(&id, NULL, signal_thread_start, NULL) != 0;
#line 6956 "APItest.c"
        sv_setpvn(ST(0), (const char *)&(id), sizeof(id));
        SvUTF8_off(ST(0));
        SvSETMAGIC(ST(0));
        TARGi((IV)RETVAL, 1);
        ST(0) = TARG;
    }
    XSRETURN(1);
}
```
The SvSETMAGIC() is generated by XS where needed.

(I have added the SvUTF8_off() to the typemap locally.)

bulk88 commented on util.c:
```diff
@@ -2872,6 +2872,18 @@ Perl_atfork_unlock(void)
+       new main thread
+    */
+    PTHREAD_INIT_SELF(PL_main_thread);
+#endif
```
More of a Win32 problem than a Posix Perl problem, but what happens to the entire Perl or 3P embedder process if the 1st Perl thread or 1st process startup thread, exits itself on Unix? Instant Process kill? If not, the TID or pthread_t obj stored in PL_main_thread now points to a zombie TID? Why would a child interp/my_perl on a 2nd pthread rethrow a signal  to a gone/finished/dealloced root my_perl and dealloced or zombie TID??
```
static void
S_init_tls_and_interp(PerlInterpreter *my_perl)
{
    if (!PL_curinterp) {
        PERL_SET_INTERP(my_perl);
#if defined(USE_ITHREADS)
        INIT_THREADS;
        ALLOC_THREAD_KEY;
        PERL_SET_THX(my_perl);
        OP_REFCNT_INIT;
        OP_CHECK_MUTEX_INIT;
        KEYWORD_PLUGIN_MUTEX_INIT;
        HINTS_REFCNT_INIT;
        LOCALE_INIT;
        USER_PROP_MUTEX_INIT;
        ENV_INIT;
        MUTEX_INIT(&PL_dollarzero_mutex);
        MUTEX_INIT(&PL_my_ctx_mutex);
        PTHREAD_INIT_SELF(PL_main_thread);
#  endif
    }
```

tonycoz replied:
Exiting the main thread exits the process, destroying all threads. 

If a parent process exits while the child process is running, the parent process exits and (typically) PID 1 becomes the parent of the child process.

tonycoz commented:
Failing on OpenBSD, looking into it https://perl5.test-smoke.org/report/5516433

tonycoz committed:
foreign thread signal forwarding: swap tests to support OpenBSD

On OpenBSD, With the tests in the original order, with the parent
process setting a signal handler and accepting a foreign signal, the
child process test would block busy waiting when the child process
called exit.

From looking at kdump the child process is busy waiting on sched_yield
outside the perl process.

Since this appears to be OpenBSD specific (other BSD, linux, MacOS
work fine), and the sequence causing the problem seems unlikely
outside a test scenario I'm not going to investigate further.

tonycoz commented:
Worked around OpenBSD failure.

jkeenan commented:
> Worked around OpenBSD failure.

Tests pass for me on OpenBSD -- but my version is the admittedly obsolete 6.9
